rfcs/20210301-computation-datatype/README.md

```diff
@@ -0,0 +1,178 @@
+
+dnnl_status_t DNNL_API dnnl_primitive_attr_get_fp_math_mode(
+        dnnl_primitive_attr_t attr, dnnl_fp_math_mode_t mode);
+```
+
+The math mode attribute will have the following values:
+```c++
+typedef enum {
+    dnnl_fp_math_mode_strict, // default behavior, described in doc
+    dnnl_fp_math_mode_bf16,   // implicit f32->bf16 conversion allowed
+    dnnl_fp_math_mode_f16,    // implicit f32->f16 conversion allowed
+    dnnl_fp_math_mode_any,    // implicit f32->f16 or f32->bf16 conversion allowed
+} dnnl_fp_math_mode_t;
+```
+
+As mentioned in the previous section, these will apply only to
+primitives created with floating-point computation. Using any value
+other value than `dnnl_fp_math_mode_strict` for primitives with
+integer computation data-type should return `invalid`. As a first
+step, we would allow implicit down-conversion only to formats that are
+at least as accurate as the format specified in the math mode (so same
+number or greater number of mantissa bits, and same or greater number
+of exponent bits). In particular, we would have this relationships:
+- `f32 > tf32 > f16`
+- and `f32 > tf32 > bf16`.
+
+Note that `bf16` and `f16` are not comparable since `bf16` has a wider
+range (more exponent bits) but less accuracy (fewer mantissa
+bits). Because `f32` is the only FP type that is comparable to all
+other FP types, we suggest to restrict the implicit down-convert
+feature only when the original format is `f32` (this should be the
+most common use-case anyway).  We could later extend it to also
+support implicit down-conversion from any FP data-type (hence why the
+enum and attribute name mentions `fp` and not `f32`).
+
+We could reuse the `dnnl_data_type_t` enum type instead of introducing
+a `dnnl_fp_math_mode_t` enum type, but I believe it is better to keep
+them separate as
+1. it does not make sense to have `dnnl_datatype_any` for data_type
+2. and we don't want to support integer types.
+
+## 3.2 New environment variable and associated function.
+
+To allow end-users to experiment with down-conversion without depending
+on FWK/toolkit knobs, we would introduce a `DNNL_DEFAULT_FP_MATH_MODE`
+environment variable. This would change the `default` value for the
+math mode attribute. The accepted values would be:
+- `STRICT` to disable implicit down-conversion. This would be the
+  default for v2.x to preserve current behavior. The default could be
+  changed in future major versions.
+- `BF16`, `F16` to allow implicit down-conversions from `f32` to
+  `bf16`/`f16` or compatible FP type.
+- `ANY` where we allow down-conversion from `f32` to any compatible
+  data-type with lower accuracy.
+
+As usual, we would introduce the corresponding function APIs, which
+supersedes the environment variable.
+
+```c++
+dnnl_status_t DNNL_API dnnl_set_default_fp_math_mode(dnnl_fp_math_mode_t mode);
+```
+
+Because some users create primitives in parallel, this setting should
+be thread local.
```

Comment:
Looks like this wouldn't be aligned with other oneDNN setters that duplicate environment variables. I think such setters have the global behavior (not thread local like the RFC suggests). Please correct me if I'm wrong.

Reply:
I agree it differs from what we have for other environment variables.
The problem is that we have users that create primitive in parallel. If we don't make the setting thread local, the user would either have to:
- use the attribute API explicitly for each primitive creation. This would typically require a change in their code for each op (so each place where they create a oneDNN primitive).
- use global setting. Here they would have to treat each primitive creation as a critical section, and that will hurt the performance of primitive creation.

Because of the overhead mentioned on bullet 2 (both in compute time, and implementation from user), I would recommend not going that route. The 2 viable options I see are:
- use thread local setting with this function API, so that user do not need to change all their primitive creation code with the new attribute, and are still able to scope the down-conversion behavior.
- make the function API a `set_once` setting. So the setting would be global to all threads, but the users would be able to change the setting only once and before any attribute/primitive creation (and return an error if they want to change it again). With this option, the scoped down-conversion would be enabled only through attributes.

Reply:
QQ: Why is it expected that the users will set the default value for each primitive creation?

Reply:
That is a good question (about the `set_compute_mode`), but it would not necessarily perfectly match a global setter in oneDNN actually. Take the following graph for example:

```
a - b - c - d
  \ e - f /
```
If the FWK user calls `set_compute_mode` only around the subgraph `e-f`, we don't want the setting to change for `b-c`. So whether thread_local matches what the FWK will do depends on how the FWK would implement such a function. If we are global only, they will have to serialize the creation of `b-c` and `e-f`. If we have a thread local setting, they will still be able to do them in parallel, but will have to create separate `buckets` of nodes that can be created in parallel.

In any case, I still think that the thread_local setting is more flexible (the user can either serialize or manage node creation context to extract more parallelism), whereas the global setting essential forces serialization.

Reply:
> If the FWK user calls set_compute_mode only around the subgraph e-f, we don't want the setting to change for b-c.

Not sure I understand. If `set_compute_mode` has global behavior in FW, users have to serialize their creation - no other choice. I agree, parallelization is possible if FW exposes thread-local behavior. Talking about TensorFlow and PyTorch, I see they have global semantics for `tf32` enabling:

- PyTorch: https://pytorch.org/docs/stable/backends.html
- TensorFlow: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/platform/tensor_float_32_utils.cc

I expect oneDNN setter would be used under similar TF/PT interfaces with global behavior. And my concern was if we go with the thread local interface we may slightly complicate FW integration as we wouldn't have one-to-one match in behavior. On the other hand if we don't have explicit requests and scenarios for now, we may start with the simplest option (global setter) and add another interface upon request (say `dnnl_set_thread_local_default_fp_math_mode`).

Reply:
The semantic of the existing knobs in python APIs (like the one you linked to for PyTorch or this [one in TF](https://www.tensorflow.org/api_docs/python/tf/config/experimental/enable_tensor_float_32_execution)) is only related to execution of ops. This works fine with eager execution but not for graph execution. In particular, it does not allow the user to scope the down-conversion as execution will be for the whole graph. My understanding is that framework and toolkit developers are interested in letting the user scope this down-conversion property during graph creation.

I don't know exactly how they will do that though. If each op node has a numerical property, the attribute API seems the most natural to use. If they plan on adding an op in the graph relative to the change of mode, then the global knob in oneDNN will not work without constraining the order of traversal of the graph inside the FWK.

With that aside, even if FWK would restrict the knob to execution time (like the current existing knobs do), what problem do you see with the thread_local behavior? In those cases, the execution of the Python script is sequential, so op creation should be sequential as well and whether the knob is thread_local or not should not matter I believe.

Reply:
> With that aside, even if FWK would restrict the knob to execution time (like the current existing knobs do), what problem do you see with the thread_local behavior? In those cases, the execution of the Python script is sequential, so op creation should be sequential as well and whether the knob is thread_local or not should not matter I believe.

Sorry, I'm not aware of how primitive creation is implemented. I assume frameworks may use a threadpool, and creation may happen in other threads. If that's the case, extra workarounds are needed to map global behavior to thread local behavior. This issue may be also applicable to scoped creation. The assumption about sequential creation looks like an implementation detail to me.

Anyway, thread-local/global behavior shouldn't be a big deal. If frameworks come across any related issues we can address them in the future.